### ISSUE TITLE
Number of icons per row in the snap panel

### ISSUE THREAD

**Issue by ReinerBforartists** [Bforartists/Bforartists#2399]
We need to sort the number of icon only buttons in the snap panel by currend width. With one row the icons are overlapping. With two rows the buttons are still squished. With three rows we have the needed width to display them all.

Maybe we can use a grid layout here. But it should be possible in the same way than with the tools buttons.

![squished2](https://user-images.githubusercontent.com/13919702/117530015-2fd32f00-afdb-11eb-9ac3-0a135b3ffe85.jpg)

![squished](https://user-images.githubusercontent.com/13919702/117530018-32358900-afdb-11eb-88db-76df1756b121.jpg)

![squished3](https://user-images.githubusercontent.com/13919702/117530020-35c91000-afdb-11eb-8789-d1e74e5f0830.jpg)

**Comment by ReinerBforartists** [comment]
This taks is to solve first: #2400 . Maybe this task becomes obsolete then.

**Comment by ReinerBforartists** [comment]
For the tools buttons this is defined in class ToolSelectPanelHelper: Pretty complicated code Maybe we can get away with a grid layout. But i doubt so. So second thought is to retreive the width_scale value, and set the icon positions that way. Would be an elseif for all three icon state cases though. Not happy.

**Comment by ReinerBforartists** [comment]
This problem here refers to the button size. We need to solve both at once. 

Resizing by grid does not take the scale_x and scale_y values into account. Mean!

Let's have a look in the Python api. Maybe there is a way to tell the flow grid to work with other values than 1.

![flowlayout](https://user-images.githubusercontent.com/13919702/117534167-99116d00-aff0-11eb-9ada-24df9a17914e.jpg)

**Comment by ReinerBforartists** [comment]
Hm, three buttons is not in the grid it seems. I can have one, two, four, six, but not three. Ah the fun. So close to the easy solution ^^

https://docs.blender.org/api/current/bpy.types.UILayout.html?highlight=grid_flow#bpy.types.UILayout.grid_flow

**Comment by ReinerBforartists** [comment]
Unassigning. Merge ahead ...

**Comment by ReinerBforartists** [comment]
A shame, it is really one, two, four with a grid_flow layout. No three. And the method for the tools is deeply connected with the whole new tool system itself. I cannot see how i can get a foot into the door here. I am not really happy when i would need to use three ifs here, with always the same content. But it seems this is the easiest solution.

**Comment by ReinerBforartists** [comment]
Ugly but working ^^

The upper four buttons are working, the buttons below not, as a reference. I end in four different layouts, with always the same buttons. This makes the code very long and unhandy. New buttons needs to be implemented four times now, each. But i don't see another way really. 

@iyadahmed, do you know a more elegant way here? As told, the ToolSelectPanelHelper that the Blender developers use for the  tools is completely tied to the tool system. I have given up at it after the tenth missing variable. I can't even proper isolate the sorting bit really, which is spread across several defines.

![workingbutugly](https://user-images.githubusercontent.com/13919702/117691246-fa396c00-b1bb-11eb-832e-74acd2b04526.jpg)

**Comment by iyadahmed** [comment]
I believe ToolSelectPanelHelper is same but auto generated.

We can keep it this way, then later we can write our own auto generation.

**Comment by ReinerBforartists** [comment]
Hm, it would be wiser to write it now. Before we throw in all panels ^^

**Comment by iyadahmed** [comment]
> Hm, it would be wiser to write it now. Before we throw in all panels ^^

agree, should it be a separate task?

**Comment by ReinerBforartists** [comment]
No, this is this task here :)

**Comment by iyadahmed** [comment]
> No, this is this task here :)

Oh ok

**Comment by ReinerBforartists** [comment]
I have dived into the possibilities a bit more. Toyed around with a panel with the new tool system. All i managed was to show the select tool twice. But there is no code to do so. This new system is so crazy ...

 Then i tested if we can at least use the new icons. But even this is too tied to the new tool system. So it remains the way it is. Unelegant, but working. Which is good enough.

Closing as fixed.

![two](https://user-images.githubusercontent.com/13919702/118085785-3e4a8d80-b3c3-11eb-8aec-09ec7c29386e.jpg)
